docs/content/sips/000-app-config.md

```diff
@@ -0,0 +1,143 @@
+title = "SIP xxx - Application Configuration"
+template = "main"
+date = "2022-03-22T14:53:30Z"
+
+---
+
+Summary: A configuration system for Spin applications.
+
+Owner: lann.martin@fermyon.com
+
+Created: March 22, 2022
+
+Updated: April 1, 2022
+
+## Background
+
+It is common for applications to require configuration at runtime that isn't known at build time or is too sensitive to be stored with build artifacts:
+
+- Logging configuration
+- Per-channel (production, staging, etc) service dependency URLs
+- Database secrets
+
+## Proposal
+
+### Configuration is defined by components and applications
+
+Configuration within a "parent" (component or application) consists of a number of configuration "slots":
+- Slots are uniquely identified within their parent by a string "key"; in order to allow unambiguous conversion to environment variables or file paths, keys are constrained:
+  - Keys must start with a letter (required for env vars)
+  - Keys consist of only lowercase ascii alphanum and `_` (`[a-z0-9_]`)
+    - Only one `_` at a time and not at the end (to allow delimiting in env vars with `__`)
+- A slot must _either_ be marked as "required" _or_ must be given a default value
+- A slot may be marked as "secret", in which case any associated value should be handled with care (e.g. not logged)
+
+```toml
+[config]
+# This simple form...
+key1 = "default_value"
+# ...is equivalent to:
+key1 = { default = "default_value" }
+# required & secret slot 
+key2 = { required = true, secret = true }
+```
+
+Default values can use template strings to reference other slots.
+```toml
+[config]
+key1 = { required = true }
+key2 = "prefix-{{ key1 }}-suffix"
+```
+
+### Components and applications can set configuration of their direct dependencies
+
+In dependency configuration, templates can reference top-level config keys (those in `[config]`), "sibling" keys within the same dependency, and "ancestor" dependant configs.
+
+- Top-level references use just the key name: `{{ top_level_key }}`
+- "Sibling" references use a single `.` prefix: `{{ .sibling_key }}`
+- "Ancestor" references use multiple `.`s: `{{ ..parent_dep_key }}`, `{{ ...grandparent_key }}`
+
+`spin.toml`:
+```toml
+[config]
+app_root = "/app"
+log_file = "{{ app_root }}/log.txt"    # -> "/app/log.txt"
+...
+[[component.config]]
+work_root = "{{ app_root }}/work"      # -> "/app/work"
+work_out = "{{ .work_root }}/output"   # -> "/app/work/output"
```

Comment:
I wonder about the value of this.  What if we had a rule that app settings propagated to components (and there were rules against shadowing)?  I suppose that would cause issues for code the app author didn't control.  But this feels a bit fiddly, like having to go edit stuff if a config setting moved from component level to app level because we started using it in a new place.

Reply:
I guess the whole hierarchy thing feels a bit unnecessary but I assume it's for components with config items with the same name but you want them to have different values?  Might be good to put a few words in about the motivation.

Reply:
Yeah, this is to support "third-party" components. I'll add words.